### DISCUSSION TITLE
How to insert hooks into 'createEditor' or how can I listen the markdown change out of 'createEditor'

### DISCUSSION THREAD

**Discussion by cruvie** [Milkdown/milkdown#648] [Q&A] [answered]
```tsx
 Editor.make().config((ctx) => {ctx.get(listenerCtx)
                .markdownUpdated((ctx, markdown) => {
                       //how to insert hooks here or any other way to user 'markdownUpdated'
                });
```

**Comment by Saul-Mirone** [comment, marked as answer]
Maybe use action.

**Reply by cruvie** [reply]
> Maybe use action.

thanks,I found the example on home page,but I don't konw what object I can listen (by userEffect) to trigger the action automotically when the content changes
```tsx

import { Editor, editorViewCtx, serializerCtx } from '@milkdown/core';

async function playWithEditor() {
    const editor = await Editor.make().use(commonmark).create();

    const getMarkdown = () =>
        editor.action((ctx) => {
            const editorView = ctx.get(editorViewCtx);
            const serializer = ctx.get(serializerCtx);
            return serializer(editorView.state.doc);
        });

    // get markdown string:
    getMarkdown();
}

```

**Reply by Saul-Mirone** [reply]
I think you should still use `listenerCtx` hook. Why don't you just pass in a function in `markdownUpdated`?

**Reply by cruvie** [reply]
> I think you should still use `listenerCtx` hook. Why don't you just pass in a function in `markdownUpdated`?

> I think you should still use `listenerCtx` hook. Why don't you just pass in a function in `markdownUpdated`?

I've already tried, I want to store every changes of the content, so I use redux-react, but in `markdownUpdated` function, redux hooks will cause some errors
```tsx

.markdownUpdated((ctx, markdown) => {
             //use redux hooks
                    const dispatch = useAppDispatch();
                   ---
                });

```

errors in browser
```

Error: Invalid hook call. Hooks can only be called inside of the body of a function component. This could happen for one of the following reasons:
1. You might have mismatching versions of React and the renderer (such as React DOM)
2. You might be breaking the Rules of Hooks
3. You might have more than one copy of React in the same app

```

**Reply by Saul-Mirone** [reply]
You should call `useAppDispatch` in your react component and pass in `dispatch` function as a parameter of the hook.

**Reply by cruvie** [reply]
> You should call `useAppDispatch` in your react component and pass in `dispatch` function as a parameter of the hook.

good job thank you🤓
